File: Data Structures & Algorithms/same-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if (not p and not q):
            return True
        return self.preOrderHelper(p, q)

    def preOrderHelper(self, x: Optional[TreeNode], y: Optional[TreeNode]) -> bool:
        if (not x and not y):
            return True
        if (not x or not y):
            return False

        if (x.val != y.val):
            return False

        return self.preOrderHelper(x.left, y.left) and self.preOrderHelper(x.right, y.right)

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        if (root == null) return 0;

        int leftDep = maxDepth(root.left);
        int rightDep = maxDepth(root.right);
        int diameter = leftDep + rightDep;

        int subTree = Math.max(diameterOfBinaryTree(root.left), diameterOfBinaryTree(root.right));
        return Math.max(diameter, subTree);
    }

    private int maxDepth(TreeNode root) {
        if (root == null) return 0;

        return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
    }
}
